File: references/betterlectio/android/app/src/main/java/dk/betterlectio/android/feature/grades/GradeModels.kt
package dk.betterlectio.android.feature.grades

/** A grade column parsed from the live KarakterGV header (order varies by school/term). */
data class GradeColumn(
    /** Canonical key used to look up values in [GradeRow.grades]. */
    val key: String,
    /** Full Danish header label from Lectio. */
    val label: String,
)

data class GradeCellValue(
    val value: String,
    val weight: Double? = null,
    val xprsSubject: String? = null,
    val source: String? = null,
)

data class GradeRow(
    val team: String,
    val subject: String,
    val teamId: String? = null,
    /** Grades keyed by [GradeColumn.key] in the order of the live table. */
    val grades: Map<String, GradeCellValue> = emptyMap(),
) {
    fun cell(columnKey: String): GradeCellValue? = grades[columnKey]

    /** Compact multi-type summary for detail headers (filled cells only). */
    val gradeSummary: String
        get() = grades.values
            .map { it.value.trim() }
            .filter { it.isNotEmpty() }
            .joinToString(" · ")
}

data class GradeNoteEntry(
    val hold: String,
    val gradeType: String,
    val grade: String,
    val insertedAt: String,
    val note: String?,
)

data class GradesReport(
    val columns: List<GradeColumn>,
    val grades: List<GradeRow>,
    val notes: List<GradeNoteEntry> = emptyList(),
    val alerts: List<String> = emptyList(),
)

data class GradeSubjectDetail(
    val row: GradeRow,
    val notes: List<GradeNoteEntry>,
    val columns: List<GradeColumn> = emptyList(),
)

File: references/betterlectio/ios/BetterLectio/GradeModels.swift
//
//  GradeModels.swift
//  BetterLectio
//
//  Created by Elliott Friedrich on 03/02/2026.
//

import Foundation

// MARK: - Grades Models

struct GradesReport: Codable, Equatable {
    let blockedWrittenProtocolTerm: String?
    let blockedOralProtocolTerm: String?
    let columns: [GradeColumn]
    let grades: [GradeEntry]
    let notes: [GradeNoteEntry]
}

/// A grade column parsed from the live Lectio table header. Schools and terms
/// do not all expose the same columns, so identity cannot safely be positional.
struct GradeColumn: Codable, Equatable, Hashable, Identifiable {
    let key: String
    let label: String

    var id: String { key }

    var shortLabel: String {
        switch key {
        case "1.standpunkt": return "1.SP"
        case "2.standpunkt": return "2.SP"
        case "3.standpunkt": return "3.SP"
        case "afsluttende": return "Afsl."
        case "intern prøve": return "Int."
        case "årskarakter": return "Års"
        case "eksamenskarakter": return "Eks."
        default:
            let compact = label.split(whereSeparator: { $0.isWhitespace }).joined(separator: " ")
            return compact.count <= 10 ? compact : String(compact.prefix(9)) + "…"
        }
    }
}

struct GradeEntry: Codable, Equatable, Identifiable {
    let id: String
    let hold: String
    let holdElementId: String?
    let subject: String
    let grades: [String: GradeCellValue]

    func cell(for columnKey: String) -> GradeCellValue? {
        grades[columnKey]
    }
}

struct GradeCellValue: Codable, Equatable {
    let value: String
    let xprsSubject: String?
    let source: String?
    let weight: Double?
}

struct GradeNoteEntry: Codable, Equatable, Identifiable {
    let id: String
    let hold: String
    let gradeType: String
    let grade: String
    let insertedAt: String
    let note: String?
}
